### PR TITLE
[wip] Investigate the cause of timeout on macOS

### LINKED ISSUE #11216: `cpu (macOS-10.15, 3.7, latest, stable)` gets cancelled due to timeout
## 🐛 Bug
The CI run always (?) fails with the following message due to timeout (set to 40 min). Similarly to #11184, the actual tests seem to always pass, but the difference is that runs here get cancelled due to timeouts instead of segfaults.

```
=== 2247 passed, 499 skipped, 11 xfailed, 3618 warnings in 613.75s (0:10:13) ===
Error: The operation was canceled.
```

### To Reproduce
- https://github.com/PyTorchLightning/pytorch-lightning/runs/4601898496?check_suite_focus=true#step:16:2860
- https://github.com/PyTorchLightning/pytorch-lightning/runs/4601604324?check_suite_focus=true#step:16:2860
- https://github.com/PyTorchLightning/pytorch-lightning/runs/4601145564?check_suite_focus=true#step:16:2860

### Expected behavior
No timeout.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0):
- PyTorch Version (e.g., 1.10):
- Python version (e.g., 3.9):
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context
Could be the same issue as #11184, but creating this issue for tracking just in case.

cc @carmocca @akihironitta @borda

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Trying to fix #11216 where `cpu (macOS-10.15, 3.7, latest, stable)` job hangs after finishing the actual testing part.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
I did a bit of bisection and found the culprits, this command reproduces the hang

```bash
python -m pytest tests/core/test_lightning_module.py::{test_toggle_untoggle_2_optimizers_no_shared_parameters,test_sharded_tensor_state_dict} tests/models/test_horovod.py::test_horovod_multi_optimizer -v
```

There is an undesired interaction between these.

And `tests/core/test_lightning_module.py::test_sharded_tensor_state_dict` fails if another test was not run before.

File: tests/models/test_horovod.py [hunk 1]
```diff
@@ -235,33 +235,9 @@ def validation_step(self, batch, *args, **kwargs):
 
 @RunIf(skip_windows=True, horovod=True)
 def test_horovod_multi_optimizer(tmpdir):
```
Comment (carmocca) [inline tests/models/test_horovod.py:237]: A single-test repro:

```python
import torch.distributed as dist

from pytorch_lightning import Trainer


def test_foo(single_process_pg):
    import torch.distributed._sharded_tensor

    spec = dist._sharding_spec.ChunkShardingSpec(dim=0, placements=["rank:0/cpu"])

    # if this is not called, it hangs during the test
    x = dist._sharded_tensor.empty(spec, 10, 20)
    # if called, it hangs on pytest teardown

    trainer = Trainer(strategy="horovod")
```

File: tests/core/test_lightning_module.py [hunk 2]
```diff
@@ -313,13 +313,33 @@ def __init__(self, spec):
 
 @RunIf(min_torch="1.10", skip_windows=True)
 def test_sharded_tensor_state_dict(tmpdir, single_process_pg):
+    # PROBLEM 1:
```
Comment (carmocca) [inline tests/core/test_lightning_module.py:316]: cc @yifuwang as the original author in #8944

also cc @ananthsub in case you know who else could help here

Comment (akihironitta) [reply tests/core/test_lightning_module.py:316]: For the note, `PROBLEM 1` should be unrelated to the hang and be fixed in #11416.

File: tests/models/test_horovod.py (hunk 1 above)
Comment (carmocca) [reply tests/models/test_horovod.py:237]: Have you been able to investigate further @akihironitta?

**Comment by akihironitta** [comment]
Seems like the issue is resolved somehow in master.
